ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,396 @@
+    return (vnodeAddress >> 3) & s_ModBitmask;
+}
+
+KEXT_STATIC bool TryGetVnodeRootFromCache(
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHashIndex,
+    uint32_t vnodeVid,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle)
+{
+    bool rootFound = false;
+    rootHandle = RootHandle_None;
+
+    RWLock_AcquireShared(s_entriesLock);
+    {
+        uintptr_t vnodeIndex;
+        if (TryFindVnodeIndex_Locked(vnode, vnodeHashIndex, /*out*/ vnodeIndex))
+        {
+            if (vnode == s_entries[vnodeIndex].vnode &&
+                vnodeVid == s_entries[vnodeIndex].vid &&
+                RootHandle_Indeterminate != s_entries[vnodeIndex].virtualizationRoot)
+            {
+                rootFound = true;
+                rootHandle = s_entries[vnodeIndex].virtualizationRoot;
+            }
+        }
+    }
+    RWLock_ReleaseShared(s_entriesLock);
+    
+    return rootFound;
+}
+
+KEXT_STATIC void FindVnodeRootFromDiskAndUpdateCache(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vfs_context_t _Nonnull context,
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHashIndex,
+    uint32_t vnodeVid,
+    UpdateCacheBehavior updateEntryBehavior,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle)
+{
+    rootHandle = VirtualizationRoot_FindForVnode(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        vnode,
+        context);
+
+    bool forceRefreshEntry;
+    VirtualizationRootHandle rootToInsert;
+    switch (updateEntryBehavior)
+    {
+      case UpdateCacheBehavior_ForceRefresh:
+        rootToInsert = rootHandle;
+        forceRefreshEntry = true;
+        break;
+        
+      case UpdateCacheBehavior_InvalidateEntry:
+        rootToInsert = RootHandle_Indeterminate;
+        forceRefreshEntry = true;
+        break;
+
+      default:
+        rootToInsert = rootHandle;
+        forceRefreshEntry = false;
+        break;
+    }
+
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        if (!TryInsertOrUpdateEntry_ExclusiveLocked(
+                vnode,
+                vnodeHashIndex,
+                vnodeVid,
+                forceRefreshEntry,
+                rootToInsert))
+        {
+            // TryInsertOrUpdateEntry_ExclusiveLocked can only fail if the cache is full
+            
+            perfTracer->IncrementCount(PrjFSPerfCounter_CacheFullCount, true /*ignoreSampling*/);
+        
+            InvalidateCache_ExclusiveLocked();
+            if(!TryInsertOrUpdateEntry_ExclusiveLocked(
+                        vnode,
+                        vnodeHashIndex,
+                        vnodeVid,
+                        true, // forceRefreshEntry
+                        rootToInsert))
+            {
+                KextLog_Error(
```

Comment:
@jeschu1 the more I think about it the more I don't like having this function excluded from the coverage report.

I'm thinking of moving this code out to a helper function so that I can force the `TryInsertOrUpdateEntry_ExclusiveLocked` to return false in the unit tests.

What are your thoughts on the approach?

I was thinking of naming the function something like `InsertEntryIntoInvalidedCache_ExclusiveLocked`.

cc: @nickgra for his thoughts as well